### ISSUE TITLE
Can't use matrix generator with git and list generator in combination of elementsyaml

### ISSUE THREAD

**Issue by SmartassSkeleton** [argoproj/argo-cd#15341]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

I am trying to create an AppSet for multiple environments using a git and a list generators as [described here](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Generators-List/#dynamically-generated-elements) via a declarative setup.

My AppSet:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  annotations:
    argocd.argoproj.io/sync-wave: "0"
  creationTimestamp: null
  name: helm-guestbook
  namespace: argocd
spec:
  generators:
  - matrix:
      generators:
      - git:
          files:
          - path: apps/helm-guestbook/config.yaml
          repoURL: ssh://git@github.com/my-repo/argocd-live.git
          requeueAfterSeconds: 20
          revision: main
      - list:
          elements: []
          elementsYaml: "{{ .key.components | toJson }}"
  syncPolicy: {}
  template:
    metadata:
      labels:
        app.kubernetes.io/managed-by: argocd
        app.kubernetes.io/name: helm-guestbook
      name: helm-guestbook-{{ .userGivenName }}
      namespace: argocd
    spec:
      destination:
        namespace: '{{ .destNamespace }}'
        server: '{{ .destServer }}'
      ignoreDifferences:
      - group: argoproj.io
        jsonPointers:
        - /status
        kind: Application
      project: helm-guestbook
      source:
        path: '{{ .srcPath }}'
        repoURL: '{{ .srcRepoURL }}'
        targetRevision: '{{ .srcTargetRevision }}'
        helm:
          releaseName: helm-guestbook-{{ .userGivenName }}
          valueFiles:
          - "apps/helm-guestbook/{{ .valuesFile }}"
      syncPolicy:
        automated:
          allowEmpty: true
          prune: true
          selfHeal: true
status: {}
```

The `apps/helm-guestbook/config.yaml` file:
```yaml
key:
  components:
    - userGivenName: hg-stg
      destNamespace: argocd-ex-stg
      destServer: "https://kubernetes.default.svc"
      srcPath: helm-guestbook
      srcRepoURL: "ssh://git@github.com/argoproj/argocd-example-apps"
      srcTargetRevision: master
      valuesFile: values.stg.yaml
    - userGivenName: hg-prod
      destNamespace: argocd-ex-prod
      destServer: "https://kubernetes.default.svc"
      srcPath: helm-guestbook
      srcRepoURL: "ssh://git@github.com/argoproj/argocd-example-apps"
      srcTargetRevision: master
      valuesFile: values.prod.yaml
```

Error:
```shell
Status:
  Conditions:
    Last Transition Time:  2023-09-04T14:06:02Z
    Message:               failed to get params for second generator in the matrix generator: child generator returned an error on parameter generation: error unmarshling decoded ElementsYaml error converting YAML to JSON: yaml: invalid map key: map[interface {}]interface {}{".key.components | toJson":interface {}(nil)}
    Reason:                ApplicationGenerationFromParamsError
    Status:                True
    Type:                  ErrorOccurred
    Last Transition Time:  2023-09-04T14:06:02Z
    Message:               failed to get params for second generator in the matrix generator: child generator returned an error on parameter generation: error unmarshling decoded ElementsYaml error converting YAML to JSON: yaml: invalid map key: map[interface {}]interface {}{".key.components | toJson":interface {}(nil)}
    Reason:                ErrorOccurred
    Status:                False
    Type:                  ParametersGenerated
    Last Transition Time:  2023-09-04T14:06:02Z
    Message:               failed to get params for second generator in the matrix generator: child generator returned an error on parameter generation: error unmarshling decoded ElementsYaml error converting YAML to JSON: yaml: invalid map key: map[interface {}]interface {}{".key.components | toJson":interface {}(nil)}
    Reason:                ApplicationGenerationFromParamsError
    Status:                False
    Type:                  ResourcesUpToDate

```

**To Reproduce**

Apply the configuration provided in the description

**Expected behavior**

The AppSet would create 2 applications based on the `config.yaml` file provided.

**Version**

```shell
argocd version                                                          
argocd: v2.8.2+unknown
  BuildDate: 2023-08-26T10:23:34Z
  GitCommit: 
  GitTreeState: 
  GitTag: 2.8.2
  GoVersion: go1.21.0
  Compiler: gc
  Platform: linux/amd64
WARN[0000] Failed to invoke grpc call. Use flag --grpc-web in grpc calls. To avoid this warning message, use flag --grpc-web. 
argocd-server: v2.8.1+2bc94af
  BuildDate: 2023-08-22T20:08:23Z
  GitCommit: 2bc94af7bd081bc4682c2c6dc005f6336c754c2b
  GitTreeState: clean
  GoVersion: go1.20.6
  Compiler: gc
  Platform: linux/amd64
  Kustomize Version: v5.1.0 2023-06-19T16:58:18Z
  Helm Version: v3.12.1+gf32a527
  Kubectl Version: v0.24.2
  Jsonnet Version: v0.20.0
```

**Comment by Yatin03** [comment]
I am also facing the same issue. While creating applicationset using argocd cli, it's giving the below error message.

argocd appset create  $ARGOCD_FILE_NAME --upsert --loglevel=debug             ✘ 0|INT at 18:14:57
WARN[0003] Failed to invoke grpc call. Use flag --grpc-web in grpc calls. To avoid this warning message, use flag --grpc-web. 
FATA[0005] rpc error: code = InvalidArgument desc = error creating ApplicationSet: ApplicationSet.argoproj.io "elementsaml12" is invalid: spec.generators[0].matrix.generators[1].list.elements: Required value

**Comment by haarchri** [comment]
this is working on my side:

```
kubectl get applicationset -n argocd      configuration-gitops-argocd  -o yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  creationTimestamp: "2023-11-13T23:30:03Z"
  generation: 1
  name: configuration-gitops-argocd
  namespace: argocd
  resourceVersion: "4754"
  uid: 12e998a1-9855-4559-8287-d348401ca794
spec:
  generators:
  - matrix:
      generators:
      - git:
          files:
          - path: configuration-gitops-argocd/infrastructure.yaml
          repoURL: https://github.com/haarchri/gitops-argocd.git
          revision: HEAD
      - list:
          elements: []
          elementsYaml: '{{ .key.components | toJson }}'
  goTemplate: true
  goTemplateOptions:
  - missingkey=error
  template:
    metadata:
      name: '{{.name}}'
    spec:
      destination:
        namespace: '{{.namespace}}'
        server: https://kubernetes.default.svc
      project: default
      sources:
      - chart: '{{.chart}}'
        helm:
          releaseName: '{{.releaseName}}'
        repoURL: '{{.repoUrl}}'
        targetRevision: '{{.version}}'
      syncPolicy:
        automated:
          selfHeal: true
        syncOptions:
        - CreateNamespace=true
status:
  conditions:
  - lastTransitionTime: "2023-11-13T23:30:32Z"
    message: Successfully generated parameters for all Applications
    reason: ApplicationSetUpToDate
    status: "False"
    type: ErrorOccurred
  - lastTransitionTime: "2023-11-13T23:30:32Z"
    message: Successfully generated parameters for all Applications
    reason: ParametersGenerated
    status: "True"
    type: ParametersGenerated
  - lastTransitionTime: "2023-11-13T23:30:32Z"
    message: ApplicationSet up to date
    reason: ApplicationSetUpToDate
    status: "True"
    type: ResourcesUpToDate
```

version: 5.51.1

**Comment by SmartassSkeleton** [comment]
Very late resolution to this, but I figured out that my problem was not having `spec.goTemplate: true` in my original manifest. This is not a bug but more of a documentation issue.

I'd requested to add a few lines about having to enable Go templates as a requirement for this to work :)

**Comment by blakepettersson** [comment]
This issue with the empty list generator should be fixed with #17424
